elastic/logstash
# give multiple pipelines all the settings

Previously, when creating pipelines from pipelines.yml,
we'd only give a pipeline the settings related to pipelines.
The PipelineSettings class was used for this.
However a pipeline may need other settings like the keystore location.

For this we instead clone the settings object and merge all the pipeline
specific settings. This is accomplished with a new method that ensures
that only pipeline level settings are overwritten in the clone.

To check that this patch works:
1. move the `config/` folder to `/tmp/`
2. create a keystore: `bin/logstash-keystore --path.settings /tmp/config create`
3. add a secret: `bin/logstash-keystore --path.settings /tmp/config add secret`
4. use that secret in a pipelines.yml config (e.g. `pipelines.yml` file):

```yml
- pipeline.id: test
  pipeline.workers: 1
  pipeline.batch.size: 1
  config.string: 'input { generator { message => "${secret}" } } filter { sleep { time => 1 } } output { stdout { } }'
```

5. run logstash with `bin/logstash --path.settings /tmp/config`
6. observe that the string doesn't get interpolated in the generator messages
7. apply this patch
8. restart logstash and see the secret :D
